DigitalLearningSolutions.Data.Tests/Services/DelegateUploadFileServiceTests.cs

```diff
@@ -0,0 +1,713 @@
+namespace DigitalLearningSolutions.Data.Tests.Services
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using ClosedXML.Excel;
+    using DigitalLearningSolutions.Data.DataServices;
+    using DigitalLearningSolutions.Data.DataServices.UserDataService;
+    using DigitalLearningSolutions.Data.Exceptions;
+    using DigitalLearningSolutions.Data.Models.DelegateUpload;
+    using DigitalLearningSolutions.Data.Models.Register;
+    using DigitalLearningSolutions.Data.Models.User;
+    using DigitalLearningSolutions.Data.Services;
+    using DigitalLearningSolutions.Data.Tests.TestHelpers;
+    using FakeItEasy;
+    using FluentAssertions;
+    using FluentAssertions.Execution;
+    using Microsoft.AspNetCore.Http;
+    using NUnit.Framework;
+
+    public class DelegateUploadFileServiceTests
+    {
+        private const int centreId = 101;
+        private IDelegateUploadFileService delegateUploadFileService = null!;
+        private IJobGroupsDataService jobGroupsDataService = null!;
+        private IRegistrationDataService registrationDataService = null!;
+        private IUserDataService userDataService = null!;
+
+        [SetUp]
+        public void SetUp()
+        {
+            jobGroupsDataService = A.Fake<IJobGroupsDataService>();
+            A.CallTo(() => jobGroupsDataService.GetJobGroupsAlphabetical()).Returns(
+                JobGroupsTestHelper.GetDefaultJobGroupsAlphabetical()
+            );
+            userDataService = A.Fake<IUserDataService>();
+            registrationDataService = A.Fake<IRegistrationDataService>();
+            delegateUploadFileService = new DelegateUploadFileService(
+                jobGroupsDataService,
+                userDataService,
+                registrationDataService
+            );
+        }
+
+        private IXLTable CreateTableFromData(IEnumerable<DelegateDataRow> data)
+        {
+            var workbook = new XLWorkbook();
+            var worksheet = workbook.AddWorksheet();
+            return worksheet.Cell(1, 1).InsertTable(data);
+        }
+
+        private DelegateDataRow SampleDelegateDataRow(

```

Comment:
private methods should usually be at the end of the file. We should also add a "Get/Create" at the beginning of the delegate data row to follow existing naming conventions.